ambry-mysql/src/main/java/com/github/ambry/accountstats/AccountStatsMySqlStore.java

```diff
@@ -310,18 +322,179 @@ public String queryRecordedMonth(String clusterName) throws SQLException {
   /**
    * Copy the row of table {@link AggregatedAccountReportsDao#AGGREGATED_ACCOUNT_REPORTS_TABLE} to {@link AggregatedAccountReportsDao#MONTHLY_AGGREGATED_ACCOUNT_REPORTS_TABLE}
    * and update the {@code monthValue} in table {@link AggregatedAccountReportsDao#AGGREGATED_ACCOUNT_REPORTS_MONTH_TABLE}.
-   * @param clusterName The clusterName
    * @param monthValue The month value.
    * @throws Exception
    */
   @Override
-  public void takeSnapshotOfAggregatedStatsAndUpdateMonth(String clusterName, String monthValue) throws Exception {
+  public synchronized void takeSnapshotOfAggregatedAccountStatsAndUpdateMonth(String monthValue) throws Exception {
     long startTimeMs = System.currentTimeMillis();
     aggregatedaccountReportsDao.copyAggregatedUsageToMonthlyAggregatedTableForCluster(clusterName);
     aggregatedaccountReportsDao.updateMonth(clusterName, monthValue);
     storeMetrics.takeSnapshotTimeMs.update(System.currentTimeMillis() - startTimeMs);
   }
 
+  @Override
+  public synchronized void storePartitionClassStats(StatsWrapper statsWrapper) throws SQLException {
+    long startTimeMs = System.currentTimeMillis();
+    // 1. Get all the partition class names in this statswrapper
+    Map<String, StatsSnapshot> partitionClassSubMap = statsWrapper.getSnapshot().getSubMap();
+    Set<String> partitionClassNames = new HashSet<>(partitionClassSubMap.keySet());
+    Map<String, Short> partitionClassNamesInDB = partitionClassReportsDao.queryPartitionClassNames(clusterName);
+    // 2. Add partition class names not in db
+    partitionClassNames.removeAll(partitionClassNamesInDB.keySet());
+    if (!partitionClassNames.isEmpty()) {
+      for (String partitionClassName : partitionClassNames) {
+        partitionClassReportsDao.insertPartitionClassName(clusterName, partitionClassName);
+      }
+      // Fresh the partition class names
+      partitionClassNamesInDB = partitionClassReportsDao.queryPartitionClassNames(clusterName);
+    }
+
+    // 3. Get partition ids under each partition class name
+    Map<String, List<Short>> partitionIdsUnderClassNames = partitionClassSubMap.entrySet()
+        .stream()
+        .collect(Collectors.toMap(Map.Entry::getKey, ent -> ent.getValue()
+            .getSubMap()
+            .keySet()
+            .stream()
+            .map(pk -> Short.valueOf(pk.substring("Partition[".length(), pk.length() - 1)))
+            .collect(Collectors.toList())));
+
+    // 4. Add partition ids not in db
+    Set<Short> partitionIdsInDB = partitionClassReportsDao.queryPartitionIds(clusterName);
+    for (String partitionClassName : partitionIdsUnderClassNames.keySet()) {
+      short partitionClassId = partitionClassNamesInDB.get(partitionClassName);
+      for (Short pid : partitionIdsUnderClassNames.get(partitionClassName)) {
+        if (!partitionIdsInDB.contains(pid)) {
+          partitionClassReportsDao.insertPartitionId(clusterName, pid, partitionClassId);
```

Comment:
Will the number of partitions ever be large enough to warrant batch insert?

Reply:
The number of partitions in each host, for now is 10 disks * 13 replicas on each disk = 130, so the batch size would be 130, which is not very big. And this is one time database write, once the partition id is written to database, we will not write it again. And last reason I didn't use a batch, is that I don't know how to deal with duplicate key in a batch operation, because other hosts, if they have other replicas of the same partition, might write to database at the same time. So we have to deal with duplicate key here.